### ISSUE TITLE
Device SDK cache inconsistency, causing device not found errors

### ISSUE THREAD

**Issue by ajcasagrande** [edgexfoundry/device-sdk-go#1331]
# 🐞 Bug Report

### Affected Services [**REQUIRED**]
device-sdk-go

### Is this a regression?
Yes

### Description and Minimal Reproduction [**REQUIRED**]
All edgex services are running normally. Called docker stop <container id of onvif service>. Wait a while. Call docker start <same container id>. Observe issue where no device commands work due to 404 device not found error, even though the devices clearly exist in core-metadata.

The first error message code is located in `device.Initialize` and it calls `for _, device := range d.sdkService.Devices() {`, which in turn calls `return cache.Devices().All()`, which leads me to believe the devices do in fact exist in the cache.

## 🔥 Exception or Error
<pre><code>
level=INFO ts=2023-02-17T16:47:20.247530253Z app=device-onvif-camera source=driver.go:141 msg="Initializing onvif client for 'tp-link-Tapo-C200-3fa1fe68-b915-4053-a3e1-1027f5ea88f4' camera"
level=ERROR ts=2023-02-17T16:47:20.440493715Z app=device-onvif-camera source=manageddevices.go:56 msg="failed to find Device tp-link-Tapo-C200-3fa1fe68-b915-4053-a3e1-1027f5ea88f4 in cache"
level=ERROR ts=2023-02-17T16:47:20.440673388Z app=device-onvif-camera source=driver.go:145 msg="failed to initialize onvif client for 'tp-link-Tapo-C200-3fa1fe68-b915-4053-a3e1-1027f5ea88f4' camera, skipping this device."
level=INFO ts=2023-02-17T16:47:20.440714682Z app=device-onvif-camera source=driver.go:141 msg="Initializing onvif client for 'HIKVISION-DS-2DE2A404IW-DE3-099f4000-4d50-11b4-82c8-c06ded544d67' camera"
level=ERROR ts=2023-02-17T16:47:20.491345631Z app=device-onvif-camera source=manageddevices.go:56 msg="failed to find Device HIKVISION-DS-2DE2A404IW-DE3-099f4000-4d50-11b4-82c8-c06ded544d67 in cache"
level=ERROR ts=2023-02-17T16:47:20.491489277Z app=device-onvif-camera source=driver.go:145 msg="failed to initialize onvif client for 'HIKVISION-DS-2DE2A404IW-DE3-099f4000-4d50-11b4-82c8-c06ded544d67' camera, skipping this device."
...
level=INFO ts=2023-02-17T16:47:23.952698178Z app=device-onvif-camera source=driver.go:610 msg="Discovered 2 device(s) in 2.44027692s via netscan."
level=INFO ts=2023-02-17T16:47:23.952931049Z app=device-onvif-camera source=async.go:104 msg="Adding discovered device HIKVISION-DS-2DE2A404IW-DE3-099f4000-4d50-11b4-82c8-c06ded544d67 to Metadata"
level=INFO ts=2023-02-17T16:47:23.958015598Z app=device-onvif-camera source=async.go:104 msg="Adding discovered device tp-link-Tapo-C200-3fa1fe68-b915-4053-a3e1-1027f5ea88f4 to Metadata"
level=ERROR ts=2023-02-17T16:47:34.382891635Z app=device-onvif-camera source=restrouter.go:168 X-Correlation-ID=8b894975-57a7-43d3-b1d6-70a83e510116 msg="device HIKVISION-DS-2DE2A404IW-DE3-099f4000-4d50-11b4-82c8-c06ded544d67 not found"
</code></pre>

## 🌍  Your Environment
**Deployment Environment:**
docker compose 2.x
edgex compose levski branch v2.3.0

**EdgeX Version [**REQUIRED**]:**
2.3.0

**Anything else relevant?**

**Comment by jinlinGuan** [comment]
I cannot recreate the error by using device-virtual or device-modbus with 2.3.0 or main branch. Could you provide us with the onvif camera setup steps, the profile and the device information?

**Comment by lenny-intel** [comment]
@ajcasagrande, is this with 2.3.0, or the latest WIP 3.0.0 code?

**Comment by presatish** [comment]
@lenny-intel facing this issue with `Levski`.

**Comment by ajcasagrande** [comment]
> @ajcasagrande, is this with 2.3.0, or the latest WIP 3.0.0 code?

Yes this is using levski branch of edgex compose, stock 2.3.0 images (ignore the first usb one):

![image](https://user-images.githubusercontent.com/48735240/220414522-3d344cdc-0ef0-4f3b-89b6-3bbab377f23a.png)

Full docker compose:
[docker-compose.yml.txt](https://github.com/edgexfoundry/device-sdk-go/files/10796040/docker-compose.yml.txt)

**Comment by ajcasagrande** [comment]
I did some debug tracing and believe I may have found the culprit. The device was being removed from cache in error, during a routine device update:

![image](https://user-images.githubusercontent.com/48735240/220424932-a8bf7cd5-ed12-4f64-a368-772abf1ca07f.png)

**Comment by ajcasagrande** [comment]
@jinlinGuan The issue is easily reproducable now. The first time a device service runs, the following line of code will correctly return the device service: 

https://github.com/edgexfoundry/device-sdk-go/blob/v2.3.0/internal/application/callback.go#L74

However, on a subsequent run, the same line of code returns an empty instance. Which will ultimately cause the following lines of code to be run, removing the device:
https://github.com/edgexfoundry/device-sdk-go/blob/v2.3.0/internal/application/callback.go#L91-L94


The reason the issue is so apparent on the device-onvif-camera is that the onvif service keeps a healthcheck of the devices and updates metadata fields about the devices. This triggers the UpdateDevice callback frequently.

To reproduce:
1. Run a clean instance of EdgeX 2.3.0
2. Add a device to a device service
3. Restart the device service
4. Update the device
    1. This should cause the device sdk to remove the device
5. Call a device command and receive a 404 device not found error

**Comment by ajcasagrande** [comment]
Ok, so I have looked into it further, and the issue is being caused due to a timing issue in regards to the actual architecture of EdgeX (device-sdk-go) itself. Since EdgeX only provides an Initialize function and not a Start function, we (device-onvif-camera devs) have placed logic in the Initialize block which will ultimately cause the UpdateDevice to be called. 

However, looking at the code:
https://github.com/edgexfoundry/device-sdk-go/blob/v2.3.0/pkg/service/init.go#L56-L67
The Initialize function is called _before_ the selfRegister function. Because of this, the dic has not been updated to reflect the current running service, thus resulting in the empty value being returned.

**Comment by lenny-intel** [comment]
@ajcasagrande, we will need to back port you hack to the levski code and release a path since this is a blocking bug, correct?

**Comment by ajcasagrande** [comment]
> @ajcasagrande, we will need to back port you hack to the levski code and release a path since this is a blocking bug, correct?

Yes, either the hack will need to be applied to the levski release of device-onvif-camera (and possibly device-usb-camera), or the alternative method would need to be patched in levski device-sdk-go.

**Comment by lenny-intel** [comment]
> Edit: Alternatively, a potential fix for the current codebase would be to move these line of code
> https://github.com/edgexfoundry/device-sdk-go/blob/v2.3.0/pkg/service/service.go#L221-L227
> to a new function which is called earlier in the bootstrap process (before driver.Initialize at least).

@ajcasagrande , this seems like a go fix for Levski and then the Start API for Minnesota for better control. Thoughts?

**Comment by ajcasagrande** [comment]
@lenny-intel Yes, I like that fix for Levski as opposed to doing the hack-y sleep method, assuming there are no other issues that arise after patching it.
